models/User.js

```diff
@@ -3,20 +3,72 @@ module.exports = (sequelize, DataTypes) => {
         username: {
             type: DataTypes.STRING,
             allowNull: false,
+            defaultValue: '', // hack to get a custom validation message for allowNull
             validate: {
-                is: ['^[a-z0-9_-]+$', 'i'], // Alphanumeric, _, -
-                len: [1, 25]
+                notEmpty: {
+                    msg: 'You must fill in a username'
+                },
+                is: {
+                    args: ['^[a-z0-9_-]+$', 'i'], // Alphanumeric, _, -
+                    msg: 'Your username can only contain alphanumeric characters, underscores and dashes'
+                },
+                len: {
+                    args: [1, 25],
+                    msg: 'Your username can only be between 1 and 25 characters'
+                },
+                isUnique: function(username, done) {
+                    // Used to improve the error message as a pre validation step instead of an index error
+
+                    User.findOne({
+                        where: sequelize.where(sequelize.fn('lower', sequelize.col('username')), username.toLowerCase()),
+                    }).then((user) => {
+                        if (!user) {
+                            done();
+                        }
+                        else {
+                            done(new Error('That username is already in use'));
+                        }
+                    }).catch((err) => {
+                        done(err);
+                    });
+                }
             }
         },
         password: {
             type: DataTypes.STRING,
-            allowNull: false
+            defaultValue: '',
+            allowNull: false,
+            validate: {
+                notEmpty: {
+                    msg: 'You must fill in a password'
+                }
+            }
         },
         email: {
             type: DataTypes.STRING,
             allowNull: true,
             validate: {
-                isEmail: true
+                isEmail: {
+                    args: true,
+                    msg: 'That email doesn\'t look correct'
```

Comment:
typo. 

Uh, is this a consistent copy and paste typo or am I crazy and it's meant to be this way? I'm starting to doubt myself. LOL.

Reply:
It escapes the `'` :)

Reply:
oooooooh. cute. <_<